RIOT-OS/RIOT
# drivers/at86rf215: implement MR-FSK

### Contribution description

This adds the MR-FSK modulation to the `at86rf215` driver.
You can configure all possible combinations of settings, grouping those settings to "FSK Modes" as per 802.15.4 is out of scope of the driver and should become part of the new 802.15.4 sub-MAC so it can be shared with other drivers implementing MR-FSK.

The same goes for channel spacing. @tinstructor added the option to toggle between 200 kHz and 400 kHz channel spacing (600 kHz is omitted for now), but again this is hardly useful isolation and should not be specific to MR-FSK.

Again, sorting this out properly would have to be done in common code (and deriving the number of channels from frequency band with / channel spacing if possible).

@jeandudey do those settings also make sense for `cc26x2_cc13x2_rf`?

### Testing procedure

You should now be able to configure the `mr-fsk` phy mode on boards with the `at86rf215` driver.

```
2020-09-07 00:37:02,866 #  ifconfig 6 set phy mr-fsk
2020-09-07 00:37:05,846 # Iface  6  HWaddr: 04:8F  Channel: 0  NID: 0x23  PHY: MR-FSK 
2020-09-07 00:37:05,854 #            modulation index: 1  2-FSK  symbol rate: 50 kHz  FEC: none  BW: 400kHz 
2020-09-07 00:37:05,858 #           Long HWaddr: 1A:F9:8F:5F:30:5D:04:8F 
2020-09-07 00:37:05,865 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-09-07 00:37:05,871 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:2022  MTU:1280  HL:64  RTR  
2020-09-07 00:37:05,873 #           6LO  IPHC  
2020-09-07 00:37:05,876 #           Source address length: 8
2020-09-07 00:37:05,879 #           Link type: wireless
2020-09-07 00:37:05,884 #           inet6 addr: fe80::18f9:8f5f:305d:48f  scope: link  VAL
2020-09-07 00:37:05,887 #           inet6 group: ff02::2
2020-09-07 00:37:05,890 #           inet6 group: ff02::1
2020-09-07 00:37:05,893 #           inet6 group: ff02::1:ff5d:48f
```

You should be able to set modulation order, symbol rate, forward error correction and modulation index.

```
2020-09-07 00:39:25,712 #  ifconfig 6 set srate 200
2020-09-07 00:39:39,103 #  ifconfig 6 set mord 4
2020-09-07 00:39:56,624 #  ifconfig 6 set fec rsc
2020-09-07 00:40:21,936 #  ifconfig 6 set modulation_index 3/4

2020-09-07 00:40:25,611 # Iface  6  HWaddr: 04:8F  Channel: 0  NID: 0x23  PHY: MR-FSK 
2020-09-07 00:40:25,619 #            modulation index: 3/4  4-FSK  symbol rate: 200 kHz  FEC: RSC  BW: 400kHz 
2020-09-07 00:40:25,623 #           Long HWaddr: 1A:F9:8F:5F:30:5D:04:8F 
2020-09-07 00:40:25,630 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-09-07 00:40:25,636 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:2022  MTU:1280  HL:64  RTR  
2020-09-07 00:40:25,638 #           6LO  IPHC  
2020-09-07 00:40:25,641 #           Source address length: 8
2020-09-07 00:40:25,644 #           Link type: wireless
2020-09-07 00:40:25,649 #           inet6 addr: fe80::18f9:8f5f:305d:48f  scope: link  VAL
2020-09-07 00:40:25,652 #           inet6 group: ff02::2
2020-09-07 00:40:25,655 #           inet6 group: ff02::1
2020-09-07 00:40:25,658 #           inet6 group: ff02::1:ff5d:48f
```

Communication should still work.
```
2020-09-07 00:43:25,820 #  ping6 ff02::1%6
2020-09-07 00:43:25,837 # 12 bytes from fe80::18f9:8f5f:305d:48f%6: icmp_seq=0 ttl=64 rssi=-30 dBm time=11.322 ms
2020-09-07 00:43:26,829 # 12 bytes from fe80::18f9:8f5f:305d:48f%6: icmp_seq=1 ttl=64 rssi=-30 dBm time=11.319 ms
2020-09-07 00:43:27,853 # 12 bytes from fe80::18f9:8f5f:305d:48f%6: icmp_seq=2 ttl=64 rssi=-30 dBm time=28.797 ms
2020-09-07 00:43:27,853 # 
2020-09-07 00:43:27,854 # --- ff02::1 PING statistics ---
2020-09-07 00:43:27,856 # 3 packets transmitted, 3 packets received, 0% packet loss
2020-09-07 00:43:27,868 # round-trip min/avg/max = 11.319/17.146/28.797 ms
```
However, not all combinations will produce valid results.


### Issues/PRs references

split off #12128
#13295 implements a driver for a different MR-FSK radio
